sway/desktop/render.c

```diff
@@ -138,18 +159,211 @@ static void render_texture(struct wlr_output *wlr_output,
+				&renderer->shaders.blur2, box, blur_radius);
+	}
+
+	pixman_region32_fini(&tempDamage);
+	pixman_region32_fini(&damage);
+
+	// Bind back to the default buffer
+	fx_framebuffer_bind(&renderer->main_buffer, width, height);
+
+	return current_buffer;
+}
+
+void render_blur(bool optimized, struct sway_output *output,
+		pixman_region32_t *output_damage, const struct wlr_fbox *src_box,
+		const struct wlr_box *dst_box, pixman_region32_t *opaque_region,
+		int surface_width, int surface_height, int32_t surface_scale,
+		struct decoration_data deco_data, struct blur_parameters blur_params) {
+	struct wlr_output *wlr_output = output->wlr_output;
+	struct fx_renderer *renderer = output->renderer;
+
+	int width, height;
+	wlr_output_transformed_resolution(output->wlr_output, &width, &height);
+	struct wlr_box monitor_box = { 0, 0, width, height };
+
+	float matrix[9];
+	enum wl_output_transform transform =
+		wlr_output_transform_invert(output->wlr_output->transform);
+	wlr_matrix_project_box(matrix, &monitor_box, transform, 0.0,
+		wlr_output->transform_matrix);
+
+	// Check if damage is inside of box rect
+	pixman_region32_t damage = create_damage(*dst_box, output_damage);
+
+	pixman_region32_t inverse_opaque;
+	pixman_region32_init(&inverse_opaque);
+	pixman_region32_t render_damage;
+	pixman_region32_init(&render_damage);
+
+	if (!pixman_region32_not_empty(&damage)) {
+		goto damage_finish;
+	}
+
+	/*
+	 * Capture the back_buffer and blur it
+	*/
+
+	pixman_box32_t surface_box = {
+		.x1 = 0,
+		.y1 = 0,
+		.x2 = dst_box->width / wlr_output->scale,
+		.y2 = dst_box->height / wlr_output->scale,
+	};
+
+	wlr_region_scale(&inverse_opaque, &inverse_opaque, wlr_output->scale);
+
+	// Gets the non opaque region
+	pixman_region32_copy(&inverse_opaque, opaque_region);
+	pixman_region32_inverse(&inverse_opaque, &inverse_opaque, &surface_box);
+	pixman_region32_intersect_rect(&inverse_opaque, &inverse_opaque, 0, 0,
+			surface_width * surface_scale,
+			surface_height * surface_scale);
+	if (!pixman_region32_not_empty(&inverse_opaque)) {
+		goto damage_finish;
+	}
+
+	wlr_region_scale(&inverse_opaque, &inverse_opaque, output->wlr_output->scale);
+
+	struct fx_framebuffer *buffer = &renderer->blur_buffer;
+	if (!buffer->texture.id || (!optimized && !config->blur_xray)) {
+		pixman_region32_translate(&inverse_opaque, dst_box->x, dst_box->y);
+		pixman_region32_intersect(&inverse_opaque, &inverse_opaque, &damage);
+
+		// Render the blur into its own buffer
+		buffer = get_main_buffer_blur(renderer, output, &inverse_opaque, matrix,
+				wlr_output->transform_matrix, dst_box, deco_data, blur_params);
+	}
+
+	/*
+	 * Draw the blurred texture
+	 */
+
+	pixman_region32_intersect_rect(&render_damage, &damage, dst_box->x, dst_box->y, dst_box->width, dst_box->height);
+	if (!pixman_region32_not_empty(&render_damage)) {
+		goto damage_finish;
+	}
+
+	// TODO: Should blur follow window opacity?
+	deco_data.alpha = 1.0;
```

Comment:
looking for some clarification here, do you mean should we blur windows that have alpha set to != 1?

Reply:
If the blurred texture should follow the surface opacity, ie if the surface opacity is set to 0.5, the blurred texture opacity should follow. But that would look kinda strange so that comment could be removed